### ISSUE TITLE
keccak: Spec and implement a sponge circuit for arbitrary-lenght inputs

### ISSUE THREAD

**Issue by CPerezz** [appliedzkp/zkevm-circuits#274]
First of all, the most important part of this issue, is to implement a multi-permutation keccak gadget which getting an input, pads it, and then applies the amount of rounds necessary in order to mix all of the chunks and digest them.

There are times when the inputs for keccak will not have a maximum bit-size. For those, we will need to figure out a way on which we don't end up with variable-lenght circuits.

_Edit_
Specs currently avaliable at https://hackmd.io/zIncjhEuRJioiIbPYfM_DQ

**Comment by barryWhiteHat** [comment]
> There are times when the inputs for keccak will not have a maximum bit-size. For those, we will need to figure out a way on which we don't end up with variable-lenght circuits.

For this i was thinking to just have the final squeese to only happen when you have absorbed the correct number of inputs . That way if you are adding more inputs you skip the squeeze and just add more inputs. Then after you squeez you start again with empty state for the next hash.

**Comment by CPerezz** [comment]
Not sure I'm following @barryWhiteHat.

So for keccak as per [spec](https://hackmd.io/sK7v0lr8Txi1bgION1rRpw?view#Overview) we basically take the state as the first `200 bytes (25 words)` and split the rest of the message as 17-word chunks which we pad.

IIUC, is your idea to add the intermediate hashes of keccak permutations for each mixing step on an entry for the lookup table of keccak? 

How I envisioned the Hashes table would look something like: 

| Input (Label as &str or String)                 | Value (Vec<Bytes>)       | Output result<Fp>                     |
| -------------------- | ------------ | ------------------------------- |
| `$Input_1_name` | `[0x0u8;355]`  |`0x08975...53453`|

So basically we just map input to output. As internally when loading the table we will need to perform all the hashes and make sure they're correct. Or some similar mechanism.

But maybe I understood wrongly what you mentioned.

**Comment by barryWhiteHat** [comment]
I was thinking to do smething like 



Input (Label as &str or String) | Value (Vec) | Output result | is_squeesed
-- | -- | --
$Input_1_name | [0x0u8;355] | 0x08975...53453 | 0 
$Input_2_name | [0x0u8;354] | 0x01231...12312 | 1

So bacailly we do the first round where we obsorb input 355. This is not the finished value so `is_squeesed = 0 `so we skip the squees and continue with what we have from the previous round in the next one. Then we absorb 354 and because is_squeese is set we finish and our final output is 0x1231..12312. 

Does this make sense. Maybe i am misunderstanding what you are saying ^ or in the spec.

**Comment by CPerezz** [comment]
I'm not sure what `squeeze` means in this context. 

Note that you need to absorb several times in this case. Specifically with an input of 355 bytes, you have 64 as the original state and then you do (355 - 64)/17 (with the padding this should be multiple of 17) mixing rounds (where you absorb).

Therefore, if you meant that you want to have each permutation round with absorb, this would require to have a table like:

__Assume padding is already done, and so from the (355-64) bytes we had originally we have (64+306) bytes after the padding  is done before starting keccak (which is multiple of 17)__

| Input (Label as &str or String)                 |   Permutation iteration    |Value (Vec<Bytes>)       | Output result<Fp>                     |
| -------------------- | ------------ | ------------------------------- |
| `$Input_1_name` | 0 |`[0x0u8;355]`  |`0x08975...53453`| (we hash the state and mix the first 64 bytes (state) and first 17 bytes of the next_input)
| `$Input_1_name` | 1 |`0x08975...53453` |`0x08898...6775`| (hash of the previous state with the next 17 inputs
| `$Input_1_name` | N |`~` |`~`| (Intemediate permutation rounds)
| `$Input_1_name` | $final_round |`0x6677....9900` |`0x03322...6766`| (we obtain the final result here)|

If this is what you meant, then it could make sense. Although the intermediate steps do not need to be in the table AFAIS, as when we check the consistency of the table the verifier will make sure to put the correct final hashes, and we save him/her to compute the intermediate steps which are more difficult to get.

So on the way I mentioned, we don't need to worry about anything except of where the constraints that ensure keccak is correct live. And this is the tricky part. 

As here, the constraints are fixed to the input lenght, but in the next round, the input can be much bigger or shorter. And therefore the ammount of constraints will differ.
Does this make sense?

**Comment by barryWhiteHat** [comment]
> If this is what you meant, then it could make sense. Although the intermediate steps do not need to be in the table AFAIS, as when we check the consistency of the table the verifier will make sure to put the correct final hashes, and we save him/her to compute the intermediate steps which are more difficult to get.

This is what i meant. I think the intermediary steps need to be in the table. Because for example if hte evm circuit is hashing a 1024 bit string. It needs to check that all of these bits are included in what the state proof hashes.

> Also, take into account that although we have a gadget for a single permutation, the goal of the Keccak circuit in theory would be to do the full mixing process and all the permutations. That's why sometimes I suggested to put a cap or something similar as otherways this becomes exploitable to insane hash requests via the keccak opcode in the EVM.

Right. So i am saying lets make the gadget adjustable. So that with a single flag it can handle just obsorb the next inputs and continue previous hash OR start a new hash. 

We will for sure set a max hash size. I think this will be defined the the gas block limit of hte zkevm at first. SO this will be the gas_block_limit / keccak_gas_cost Is this enough to avoid the concern?

**Comment by CPerezz** [comment]
>We will for sure set a max hash size. I think this will be defined the the gas block limit of hte zkevm at first. SO this will be the gas_block_limit / keccak_gas_cost Is this enough to avoid the concern?

It is. I think by doing this though anyway, we should probably filter this away in the bus-mapping and simply not transfer the Tx to the circuit builder. Otherways, we will panic for the entire block, and the witnesses as the `gc` will need to be re-computed. Which we could definitely avoid filtering out earlier than in the circuit builder.

>This is what i meant. I think the intermediary steps need to be in the table. Because for example if hte evm circuit is hashing a 1024 bit string. It needs to check that all of these bits are included in what the state proof hashes.

So on the way the `KeccakF` hash gadget is build, it forces the copy of the cells between it's internal steps and between permutations.
That means that there's no way to cheat on the intermediate steps as the copy constraints will catch the error. So taking that into account, if the output matches what is in the table, the keccak execution is already checking that all the intermediate steps are linked and therefore, they don't need to appear anywhere.
Maybe I missed something. But we thought that the copy_constraints would avoid that much work in table level.

**Comment by CPerezz** [comment]
Specs currently avaliable at: https://hackmd.io/zIncjhEuRJioiIbPYfM_DQ

**Comment by CPerezz** [comment]
Closed via https://github.com/appliedzkp/zkevm-specs/pull/161 :tada:
